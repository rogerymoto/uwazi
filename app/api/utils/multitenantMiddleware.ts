import { Request, Response, NextFunction } from 'express';
import { tenants } from 'api/odm/tenantContext';

const multitenantMiddleware = (req: Request, _res: Response, next: NextFunction) => {
  tenants
    .run(async () => {
      next();
    }, req.get('tenant'))
    .catch(next);
};

export { multitenantMiddleware };
