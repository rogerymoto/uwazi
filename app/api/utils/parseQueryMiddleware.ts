import { Request, Response, NextFunction } from 'express';

interface query {
  [key: string]: string;
}
const parseQueryProperty = (query: query, property: string) => {
  try {
    return JSON.parse(query[property]);
  } catch (e) {
    return query[property];
  }
};

export const parseQuery = (req: Request, _res: Response, next: NextFunction) => {
  req.query = Object.keys(req.query).reduce((parsedQuery: query, key: string) => {
    parsedQuery[key] = parseQueryProperty(req.query, key);
    return parsedQuery;
  }, {});

  next();
};
