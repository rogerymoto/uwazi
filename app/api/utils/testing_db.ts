import mongoose from 'mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { Db, ObjectId } from 'mongodb';
import { FileType } from 'shared/types/fileType';
import { EntitySchema } from 'shared/types/entityType';

mongoose.Promise = Promise;
mongoose.set('useFindAndModify', false);

let connected = false;
let mongod: MongoMemoryServer;
let mongodb: Db;

export type DBFixture = {
  files?: FileType[];
  entities?: EntitySchema[];
  [k: string]: any;
};

const fixturer = {
  async clear(db: Db, _collections: string[] | undefined = undefined): Promise<void> {
    const collections: string[] =
      _collections || (await db.listCollections().toArray()).map(c => c.name);

    await Promise.all(
      collections.map(async c => {
        await db.collection(c).deleteMany({});
      })
    );
  },

  async clearAllAndLoad(db: Db, fixtures: DBFixture) {
    await this.clear(db);
    await Promise.all(
      Object.keys(fixtures).map(async collectionName => {
        await db.collection(collectionName).insertMany(fixtures[collectionName]);
      })
    );
  },
};

const initMongoServer = async () => {
  mongod = new MongoMemoryServer();
  const uri = await mongod.getUri();
  await mongoose.connect(uri, {
    useUnifiedTopology: true,
    useNewUrlParser: true,
    useCreateIndex: true,
  });
  connected = true;
};

const testingDB: {
  mongodb: Db | null;
  connect: () => Promise<void>;
  disconnect: () => Promise<void>;
  id: (id?: string | undefined) => ObjectId;
  clear: (collections?: string[] | undefined) => Promise<void>;
  clearAllAndLoad: (fixtures: DBFixture) => Promise<void>;
} = {
  mongodb: null,

  async connect() {
    if (!connected) {
      await initMongoServer();
      mongodb = mongoose.connections[0].db;
      this.mongodb = mongodb;
    }
  },

  async disconnect() {
    await mongoose.disconnect();
    if (mongod) {
      await mongod.stop();
    }
  },

  id(id = undefined) {
    return mongoose.Types.ObjectId(id);
  },

  async clear(collections = undefined) {
    await fixturer.clear(mongodb, collections);
  },

  async clearAllAndLoad(fixtures: DBFixture) {
    await this.connect();
    await fixturer.clearAllAndLoad(mongodb, fixtures);
  },
};

export default testingDB;
