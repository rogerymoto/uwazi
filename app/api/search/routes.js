import Joi from 'joi';
import entities from 'api/entities';
import search from './search';
import { validation } from '../utils';
import needsAuthorization from '../auth/authMiddleware';

const parseQueryProperty = (query, property) => {
  try {
    return JSON.parse(query[property]);
  } catch (e) {
    return query[property];
  }
};

export default app => {
  app.get(
    '/api/search/count_by_template',
    validation.validateRequest(
      Joi.object()
        .keys({
          templateId: Joi.string().required(),
        })
        .required(),
      'query'
    ),
    (req, res, next) =>
      entities
        .countByTemplate(req.query.templateId)
        .then(results => res.json(results))
        .catch(next)
  );

  app.get(
    '/api/search',
    (req, _res, next) => {
      console.log(req.query);

      req.query = Object.keys(req.query).reduce((parsedQuery, key) => {
        parsedQuery[key] = parseQueryProperty(req.query, key);
        return parsedQuery;
      }, {});

      next();
    },
    validation.validateRequest({
      properties: {
        query: {
          properties: {
            filters: { type: 'object' },
            types: { type: 'array', items: [{ type: 'string' }] },
            _types: { type: 'array', items: [{ type: 'string' }] },
            fields: { type: 'string' },
            allAggregations: { type: 'boolean' },
            userSelectedSorting: { type: 'string' },
            aggregations: { type: 'string' },
            order: { type: 'string', enum: ['asc', 'desc'] },
            sort: { type: 'string' },
            limit: { type: 'number' },
            offset: { type: 'number' },
            searchTerm: { type: 'string' },
            includeUnpublished: { type: 'boolean' },
            treatAs: { type: 'string' },
            unpublished: { type: 'boolean' },
            select: { type: 'array', items: [{ type: 'string' }] },
            geolocation: { type: 'boolean' },
          },
        },
      },
    }),

    (req, res, next) => {
      const action = req.query.geolocation ? 'searchGeolocations' : 'search';

      return search[action](req.query, req.language, req.user)
        .then(results => res.json(results))
        .catch(next);
    }
  );

  app.get(
    '/api/search_snippets',
    validation.validateRequest(
      Joi.object().keys({
        searchTerm: Joi.string().allow(''),
        id: Joi.string(),
      }),
      'query'
    ),
    (req, res, next) =>
      search
        .searchSnippets(req.query.searchTerm, req.query.id, req.language)
        .then(results => res.json(results))
        .catch(next)
  );

  app.get('/api/search/unpublished', needsAuthorization(['admin', 'editor']), (req, res, next) => {
    search
      .getUploadsByUser(req.user, req.language)
      .then(response => res.json({ rows: response }))
      .catch(next);
  });
};
