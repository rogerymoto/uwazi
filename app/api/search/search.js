/* eslint-disable max-lines */
import date from 'api/utils/date';

import propertiesHelper from 'shared/comonProperties';

import translate, { getLocaleTranslation, getContext } from 'shared/translate';
import translations from 'api/i18n/translations';
import elasticIndexes from 'api/config/elasticIndexes';

import dictionariesModel from 'api/thesauri/dictionariesModel';
import { createError } from 'api/utils';
import relationtypes from 'api/relationtypes';
import documentQueryBuilder from './documentQueryBuilder';
import elastic from './elastic';
import entities from '../entities';
import templatesModel from '../templates';
import { bulkIndex, indexEntities } from './entitiesIndex';

function processFilters(filters, properties) {
  return Object.keys(filters || {}).reduce((res, filterName) => {
    const suggested = filterName.startsWith('__');
    const propertyName = suggested ? filterName.substring(2) : filterName;
    const property = properties.find(p => p.name === propertyName);
    if (!property) {
      return res;
    }

    let { type } = property;
    const value = filters[filterName];
    if (['date', 'multidate', 'numeric'].includes(property.type)) {
      type = 'range';
    }
    if (['select', 'multiselect', 'relationship'].includes(property.type)) {
      type = 'multiselect';
    }
    if (property.type === 'multidaterange' || property.type === 'daterange') {
      type = 'daterange';
    }

    if (['multidaterange', 'daterange', 'date', 'multidate'].includes(property.type)) {
      value.from = date.descriptionToTimestamp(value.from);
      value.to = date.descriptionToTimestamp(value.to);
    }

    if (property.type === 'relationshipfilter') {
      return [
        ...res,
        {
          ...property,
          value,
          suggested,
          type,
          filters: property.filters.map(f => ({
            ...f,
            name: `${f.name}.value`,
          })),
        },
      ];
    }

    return [
      ...res,
      {
        ...property,
        value,
        suggested,
        type,
        name: `${property.name}.value`,
      },
    ];
  }, []);
}

function agregationProperties(properties) {
  return properties
    .filter(
      property =>
        property.type === 'select' ||
        property.type === 'multiselect' ||
        property.type === 'relationship' ||
        property.type === 'relationshipfilter' ||
        property.type === 'nested'
    )
    .map(property => {
      if (property.type === 'relationshipfilter') {
        return {
          ...property,
          filters: property.filters.map(f => ({ ...f, name: `${f.name}.value` })),
        };
      }
      return { ...property, name: `${property.name}.value` };
    });
}

function metadataSnippetsFromSearchHit(hit) {
  const defaultSnippets = { count: 0, snippets: [] };
  if (hit.highlight) {
    const metadataHighlights = hit.highlight;
    const metadataSnippets = Object.keys(metadataHighlights).reduce((foundSnippets, field) => {
      const fieldSnippets = { field, texts: metadataHighlights[field] };
      return {
        count: foundSnippets.count + fieldSnippets.texts.length,
        snippets: [...foundSnippets.snippets, fieldSnippets],
      };
    }, defaultSnippets);
    return metadataSnippets;
  }
  return defaultSnippets;
}

function fullTextSnippetsFromSearchHit(hit) {
  if (hit.inner_hits && hit.inner_hits.fullText.hits.hits.length) {
    const regex = /\[\[(\d+)\]\]/g;

    const fullTextHighlights = hit.inner_hits.fullText.hits.hits[0].highlight;
    const fullTextLanguageKey = Object.keys(fullTextHighlights)[0];
    const fullTextSnippets = fullTextHighlights[fullTextLanguageKey].map(snippet => {
      const matches = regex.exec(snippet);
      return {
        text: snippet.replace(regex, ''),
        page: matches ? Number(matches[1]) : 0,
      };
    });
    return fullTextSnippets;
  }
  return [];
}

function snippetsFromSearchHit(hit) {
  const snippets = {
    count: 0,
    metadata: [],
    fullText: [],
  };

  const metadataSnippets = metadataSnippetsFromSearchHit(hit);
  const fullTextSnippets = fullTextSnippetsFromSearchHit(hit);
  snippets.count = metadataSnippets.count + fullTextSnippets.length;
  snippets.metadata = metadataSnippets.snippets;
  snippets.fullText = fullTextSnippets;

  return snippets;
}

function searchGeolocation(documentsQuery, templates) {
  documentsQuery.limit(9999);
  documentsQuery.offset(0);
  const geolocationProperties = [];

  templates.forEach(template => {
    template.properties.forEach(prop => {
      if (prop.type === 'geolocation') {
        geolocationProperties.push(`${prop.name}`);
      }

      if (prop.type === 'relationship' && prop.inherit) {
        const contentTemplate = templates.find(t => t._id.toString() === prop.content.toString());
        const inheritedProperty = contentTemplate.properties.find(
          p => p._id.toString() === prop.inheritProperty.toString()
        );
        if (inheritedProperty.type === 'geolocation') {
          geolocationProperties.push(`${prop.name}`);
        }
      }
    });
  });

  documentsQuery.hasMetadataProperties(geolocationProperties);

  const selectProps = geolocationProperties
    .map(p => `metadata.${p}`)
    .concat(['title', 'template', 'sharedId', 'language']);

  documentsQuery.select(selectProps);
}

function processResponse(filters, response) {
  const rows = response.hits.hits.map(hit => {
    const result = hit._source;
    result._explanation = hit._explanation;
    result.snippets = snippetsFromSearchHit(hit);
    result._id = hit._id;
    return result;
  });

  Object.keys(response.aggregations.all).forEach(aggregationKey => {
    const aggregation = response.aggregations.all[aggregationKey];
    if (aggregation.buckets && !Array.isArray(aggregation.buckets)) {
      aggregation.buckets = Object.keys(aggregation.buckets).map(key =>
        Object.assign({ key }, aggregation.buckets[key])
      );
    }
    if (aggregation.buckets) {
      const missingBucket = aggregation.buckets.find(b => b.key === 'missing');
      const keyFilters = ((filters || {})[aggregationKey.replace('.value', '')] || {}).values || [];
      const filterNoneOrMissing =
        !keyFilters.filter(v => v !== 'any').length || keyFilters.find(v => v === 'missing');
      if (aggregationKey !== '_types') {
        const anyCount =
          (typeof response.hits.total === 'object'
            ? response.hits.total.value
            : response.hits.total) -
          (missingBucket && filterNoneOrMissing ? missingBucket.filtered.doc_count : 0);
        aggregation.buckets.push({
          key: 'any',
          doc_count: anyCount,
          filtered: { doc_count: anyCount },
        });
      }
      response.aggregations.all[aggregationKey] = aggregation;
    }
    if (!aggregation.buckets) {
      Object.keys(aggregation).forEach(key => {
        if (aggregation[key].buckets) {
          const buckets = aggregation[key].buckets.map(option =>
            Object.assign({ key: option.key }, option.filtered.total)
          );
          response.aggregations.all[key] = {
            doc_count: aggregation[key].doc_count,
            buckets,
          };
        }
      });
    }
  });

  response.aggregations.all = Object.keys(response.aggregations.all).reduce(
    (allAgregations, key) => {
      allAgregations[key.replace('.value', '')] = response.aggregations.all[key];
      return allAgregations;
    },
    {}
  );
  return { rows, totalRows: response.hits.total.value, aggregations: response.aggregations };
}

const determineInheritedProperties = templates =>
  templates.reduce((memo, template) => {
    const inheritedProperties = memo;
    template.properties.forEach(property => {
      if (property.type === 'relationship' && property.inherit) {
        const contentTemplate = templates.find(
          t => t._id.toString() === property.content.toString()
        );
        const inheritedProperty = contentTemplate.properties.find(
          p => p.type === 'geolocation' && p._id.toString() === property.inheritProperty.toString()
        );
        if (inheritedProperty) {
          inheritedProperties[template._id.toString()] =
            inheritedProperties[template._id.toString()] || {};
          inheritedProperties[template._id.toString()][property.name] = {
            base: property,
            target: inheritedProperty,
          };
        }
      }
    });
    return inheritedProperties;
  }, {});

const whatToFetchByTemplate = (baseResults, templatesInheritedProperties) => {
  const toFetchByTemplate = {};

  baseResults.rows.forEach(row => {
    Object.keys(row.metadata || {}).forEach(name => {
      if (Object.keys(templatesInheritedProperties[row.template.toString()] || []).includes(name)) {
        const inheritedProperty = templatesInheritedProperties[row.template.toString()][name];
        const template = inheritedProperty.base.content;
        toFetchByTemplate[template] = toFetchByTemplate[template] || {
          entities: [],
          properties: [],
        };
        toFetchByTemplate[template].entities = toFetchByTemplate[template].entities.concat(
          row.metadata[name].map(mo => mo.value)
        );
        if (!toFetchByTemplate[template].properties.includes(inheritedProperty.target.name)) {
          toFetchByTemplate[template].properties.push(inheritedProperty.target.name);
        }
      }
    });
  });

  return toFetchByTemplate;
};

const getInheritedEntitiesData = async (toFetchByTemplate, language, user) =>
  Promise.all(
    Object.keys(toFetchByTemplate).map(t => {
      const query = { language, sharedId: { $in: toFetchByTemplate[t].entities } };
      if (!user) {
        query.published = true;
      }
      return entities.get(query, {
        ...toFetchByTemplate[t].properties.reduce(
          (memo, n) => Object.assign(memo, { [`metadata.${n}.value`]: 1 }),
          {}
        ),
        sharedId: 1,
      });
    })
  );

const getInheritedEntities = async (results, language, user) => {
  const templates = await templatesModel.get();
  const templatesInheritedProperties = determineInheritedProperties(templates);
  const toFetchByTemplate = whatToFetchByTemplate(results, templatesInheritedProperties);
  const inheritedEntitiesData = await getInheritedEntitiesData(toFetchByTemplate, language, user);

  const inheritedEntities = inheritedEntitiesData.reduce((_memo, templateEntities) => {
    const memo = _memo;
    templateEntities.forEach(e => {
      memo[e.sharedId] = e;
    });
    return memo;
  }, {});

  return { templatesInheritedProperties, inheritedEntities };
};

const entityHasGeolocation = entity =>
  entity.metadata &&
  !!Object.keys(entity.metadata)
    .filter(field => entity.metadata[field])
    .find(field => {
      if (/_geolocation/.test(field) && entity.metadata[field].length) {
        return true;
      }
      if (Array.isArray(entity.metadata[field])) {
        return !!entity.metadata[field].find(
          f => f.inherit_geolocation && f.inherit_geolocation.length
        );
      }
      return false;
    });

const processGeolocationResults = (_results, templatesInheritedProperties, inheritedEntities) => {
  const results = _results;
  const processedRows = [];
  const affectedTemplates = Object.keys(templatesInheritedProperties);

  results.rows.forEach(_row => {
    const row = _row;
    if (affectedTemplates.includes(row.template)) {
      Object.keys(row.metadata).forEach(property => {
        if (templatesInheritedProperties[row.template][property]) {
          row.metadata[property].forEach(({ value: entity }, index) => {
            const targetProperty = templatesInheritedProperties[row.template][property].target.name;
            const inherited = inheritedEntities[entity]
              ? inheritedEntities[entity]
              : { metadata: {} };
            inherited.metadata = inherited.metadata || {};
            row.metadata[property][index] = {
              value: entity,
              // geolocation may not be sanitized...
              inherit_geolocation: (inherited.metadata[targetProperty] || []).filter(p => p.value),
            };
          });
        }
      });
    }
    if (entityHasGeolocation(row)) {
      processedRows.push(row);
    }
  });

  results.rows = processedRows;
  results.totalRows = processedRows.length;
  return results;
};

const escapeCharByRegex = (query, regex, char) => {
  let escaped = query;

  regex.lastIndex = 0;
  let result = regex.exec(escaped);
  while (result) {
    const firstPart = escaped.substr(0, result.index);
    const charCount = result[1] ? 2 : 1;
    const secondPart = escaped.substr(result.index + charCount, query.length);
    const replacement = `${charCount === 2 ? result[1] : ''}\\${char}`;

    escaped = firstPart + replacement + secondPart;
    regex.lastIndex = 0;
    result = regex.exec(escaped);
  }

  return escaped;
};

const checkCharParity = (query, regex) => {
  const indices = [];

  regex.lastIndex = 0;
  let result = regex.exec(query);
  while (result) {
    indices.push(result.index);
    result = regex.exec(query);
  }

  return !(indices.length % 2 === 1);
};

const escapeElasticSearchQueryString = query => {
  const charsRegex = {
    '"': /([^\\])"|^"/g,
    '/': /([^\\])\/|^\//g,
  };

  return Object.keys(charsRegex).reduce((escaped, char) => {
    if (!checkCharParity(escaped, charsRegex[char])) {
      return escapeCharByRegex(escaped, charsRegex[char], char);
    }
    return escaped;
  }, query);
};

const instanceSearch = elasticIndex => ({
  search(query, language, user) {
    return Promise.all([
      templatesModel.get(),
      dictionariesModel.get(),
      relationtypes.get(),
      translations.get(),
    ]).then(([templates, dictionaries, relationTypes, _translations]) => {
      const textFieldsToSearch =
        query.fields ||
        propertiesHelper
          .allUniqueProperties(templates)
          .map(prop =>
            ['text', 'markdown'].includes(prop.type)
              ? `metadata.${prop.name}.value`
              : `metadata.${prop.name}.label`
          )
          .concat(['title', 'fullText']);
      const elasticSearchTerm =
        query.searchTerm && escapeElasticSearchQueryString(query.searchTerm);
      const documentsQuery = documentQueryBuilder()
        .fullTextSearch(elasticSearchTerm, textFieldsToSearch, 2)
        .filterByTemplate(query.types)
        .filterById(query.ids)
        .language(language);

      if (query.from) {
        documentsQuery.from(query.from);
      }

      if (query.limit) {
        documentsQuery.limit(query.limit);
      }

      if (query.offset) {
        documentsQuery.offset(query.offset);
      }

      if (query.includeUnpublished && user && !query.unpublished) {
        documentsQuery.includeUnpublished();
      }

      if (query.unpublished && user) {
        documentsQuery.unpublished();
      }

      const allTemplates = templates.map(t => t._id);
      const allUniqueProps = propertiesHelper.allUniqueProperties(templates);
      const filteringTypes = query.types && query.types.length ? query.types : allTemplates;
      let properties = propertiesHelper.comonFilters(templates, relationTypes, filteringTypes);
      properties =
        !query.types || !query.types.length
          ? propertiesHelper.defaultFilters(templates)
          : properties;

      if (query.sort) {
        const sortingProp = allUniqueProps.find(p => `metadata.${p.name}` === query.sort);
        if (sortingProp && sortingProp.type === 'select') {
          const dictionary = dictionaries.find(d => d._id.toString() === sortingProp.content);
          const translation = getLocaleTranslation(_translations, language);
          const context = getContext(translation, dictionary._id.toString());
          const keys = dictionary.values.reduce((result, value) => {
            result[value.id] = translate(context, value.label, value.label);
            return result;
          }, {});
          documentsQuery.sortByForeignKey(query.sort, keys, query.order);
        } else {
          documentsQuery.sort(query.sort, query.order);
        }
      }

      if (query.allAggregations) {
        properties = allUniqueProps;
      }

      // this is where we decide which aggregations to send to elastic
      const aggregations = agregationProperties(properties);
      const filters = processFilters(query.filters, [...allUniqueProps, ...properties]);
      // this is where the query filters are built
      documentsQuery.filterMetadata(filters);
      // this is where the query aggregations are built
      documentsQuery.aggregations(aggregations, dictionaries);
      if (query.select) {
        documentsQuery.select(query.select);
      }

      if (query.geolocation) {
        searchGeolocation(documentsQuery, templates);
      }

      // documentsQuery.query() is the actual call
      return elastic
        .search({ index: elasticIndex || elasticIndexes.index, body: documentsQuery.query() })
        .then(response => processResponse(query.filters, response))
        .catch(e => {
          throw createError(e.message, 400);
        });
    });
  },

  async searchGeolocations(query, language, user) {
    let results = await this.search({ ...query, geolocation: true }, language, user);

    if (results.rows.length) {
      const { templatesInheritedProperties, inheritedEntities } = await getInheritedEntities(
        results,
        language,
        user
      );
      results = processGeolocationResults(results, templatesInheritedProperties, inheritedEntities);
    }

    return results;
  },

  getUploadsByUser(user, language) {
    return entities.get({ user: user._id, language, published: false });
  },

  async searchSnippets(searchTerm, sharedId, language) {
    const templates = await templatesModel.get();

    const elasticSearchTerm = searchTerm && escapeElasticSearchQueryString(searchTerm);

    const searchFields = propertiesHelper
      .textFields(templates)
      .map(prop => `metadata.${prop.name}.value`)
      .concat(['title', 'fullText']);
    const query = documentQueryBuilder()
      .fullTextSearch(elasticSearchTerm, searchFields, 9999)
      .includeUnpublished()
      .filterById(sharedId)
      .language(language)
      .query();

    const response = await elastic.search({
      index: elasticIndex || elasticIndexes.index,
      body: query,
    });
    if (response.hits.hits.length === 0) {
      return {
        count: 0,
        metadata: [],
        fullText: [],
      };
    }
    return snippetsFromSearchHit(response.hits.hits[0]);
  },

  async indexEntities(query, select = '', limit = 200, batchCallback = () => {}) {
    return indexEntities(query, select, limit, {
      batchCallback,
      elasticIndex: elasticIndex || elasticIndexes.index,
      searchInstance: this,
    });
  },

  async bulkIndex(docs, action = 'index', index = elasticIndex) {
    return bulkIndex(docs, action, index);
  },

  bulkDelete(docs) {
    const body = docs.map(doc => ({
      delete: { _index: elasticIndex || elasticIndexes.index, _id: doc._id },
    }));
    return elastic.bulk({ body });
  },

  delete(entity) {
    const id = entity._id.toString();
    return elastic.delete({ index: elasticIndex || elasticIndexes.index, id });
  },

  deleteLanguage(language) {
    const query = { query: { match: { language } } };
    return elastic.deleteByQuery({ index: elasticIndex || elasticIndexes.index, body: query });
  },
});

export { instanceSearch };

export default instanceSearch();
