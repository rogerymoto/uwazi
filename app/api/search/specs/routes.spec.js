import entities from 'api/entities';
import { catchErrors } from 'api/utils/jasmineHelpers';
import documentRoutes from '../routes.js';
import instrumentRoutes from '../../utils/instrumentRoutes';
import search from '../search';

describe('search routes', () => {
  let routes;

  beforeEach(() => {
    routes = instrumentRoutes(documentRoutes);
  });

  describe('/api/search/count_by_template', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/search/count_by_template')).toMatchSnapshot();
    });

    it('should return count of search using a specific template', done => {
      spyOn(entities, 'countByTemplate').and.returnValue(new Promise(resolve => resolve(2)));
      const req = { query: { templateId: 'templateId' } };

      routes
        .get('/api/search/count_by_template', req)
        .then(response => {
          expect(response).toEqual(2);
          done();
        })
        .catch(catchErrors(done));
    });
  });

  describe('/api/search', () => {
    beforeEach(() => {
      spyOn(search, 'search').and.returnValue(new Promise(resolve => resolve('results')));
      spyOn(search, 'searchGeolocations').and.returnValue(
        new Promise(resolve => resolve('geolocation results'))
      );
    });

    const assessSearch = async (req, action, expectedResults, expectedArgs) => {
      const response = await routes.get('/api/search', req);
      expect(search[action]).toHaveBeenCalledWith(...expectedArgs);
      expect(response).toEqual(expectedResults);
    };

    it('should search search and return the results', async () => {
      const filtersValue = { property: 'property' };
      const types = ['ruling', 'judgement'];
      const fields = ['field1', 'field2'];
      const req = {
        query: { searchTerm: 'test', filters: filtersValue, types, fields },
        language: 'es',
        user: 'user',
      };

      const expectedArgs = [
        {
          searchTerm: 'test',
          filters: { property: 'property' },
          types: ['ruling', 'judgement'],
          fields: ['field1', 'field2'],
        },
        'es',
        'user',
      ];
      await assessSearch(req, 'search', 'results', expectedArgs);
    });

    describe('when has no filters or types', () => {
      it('should search search and return the results', async () => {
        const req = { query: {}, language: 'es', user: 'user' };
        await assessSearch(req, 'search', 'results', [{}, 'es', 'user']);
      });
    });

    describe('geolocation search', () => {
      it('should point to searchGeolocations', async () => {
        const req = { query: { geolocation: true }, language: 'es', user: 'user' };
        await assessSearch(req, 'searchGeolocations', 'geolocation results', [
          { geolocation: true },
          'es',
          'user',
        ]);
      });
    });
  });

  describe('/api/search_snippets', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/search_snippets')).toMatchSnapshot();
    });

    it('should search', done => {
      spyOn(search, 'searchSnippets').and.returnValue(new Promise(resolve => resolve('results')));
      const req = { query: { searchTerm: 'test', id: 'id' }, language: 'es' };

      routes
        .get('/api/search_snippets', req)
        .then(response => {
          expect(response).toEqual('results');
          expect(search.searchSnippets).toHaveBeenCalledWith('test', 'id', 'es');
          done();
        })
        .catch(catchErrors(done));
    });
  });

  describe('/api/search/unpublished', () => {
    it('should search', done => {
      spyOn(search, 'getUploadsByUser').and.returnValue(new Promise(resolve => resolve('results')));
      const req = { query: { searchTerm: 'test', id: 'id' }, language: 'es' };

      routes
        .get('/api/search/unpublished', req)
        .then(response => {
          expect(response).toEqual({ rows: 'results' });
          done();
        })
        .catch(catchErrors(done));
    });
  });
});
