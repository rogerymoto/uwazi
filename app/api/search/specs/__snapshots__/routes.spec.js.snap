// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search routes /api/search/count_by_template should have a validation schema 1`] = `
Object {
  "children": Object {
    "templateId": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`search routes /api/search_snippets should have a validation schema 1`] = `
Object {
  "children": Object {
    "id": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "searchTerm": Object {
      "type": "string",
      "valids": Array [
        "",
      ],
    },
  },
  "type": "object",
}
`;
